src/cider/exec/operator/join/HashTableFactory.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <map>
+#include <memory>
+#include <string>
+
+namespace cider_hashtable {
+
+// to be added
+enum hashtableName {
+  LINEAR_PROBING_INT,
+  LINEAR_PROBING_BATCH,
+  CHAINED_INT,
+  CHAINED_BATCH
+};
+
+template <class HashTableType_t>
+class IHashTableRegistrar {
+ public:
+  virtual std::unique_ptr<HashTableType_t> createHashTable() = 0;
+
+ protected:
+  IHashTableRegistrar() {}
+  virtual ~IHashTableRegistrar() {}
+
+ private:
+  IHashTableRegistrar(const IHashTableRegistrar&);
+  const IHashTableRegistrar& operator=(const IHashTableRegistrar&);
+};
+
+template <class HashTableType_t>
+class HashTableFactory {
```

Comment:
The purpose of HashTableFactory is used to create a  hashtable intance for either  join or aggregate,  may be we only need two methods , one for join and another for agg just like the velox does.

I can't see any benefit that you use Registrar to registrar a hashTable before use it.

Reply:
The purpose of HashTableFactory is to hide all hashtable's constructor into this class. The main difference for our hashtable and velox's hashtable is that we have many and they have only one. 
I guess you are thinking making this class looks like this
```
template <typename Key,
          typename Value,
          typename Hash,
          typename KeyEqual,
          typename Grower,
          typename Allocator>
class hashtablefactory 

template <typename... Args>
BaseHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator> createForJoin (HastTableType hashtable_type, Args.. args) {
  switch(hashtable_type)
      case xxx:
          return new xx_hashtable<Key, Value, Hash, KeyEqual, Grower, Allocator>(args);
      case yyy:
          return new yy_hashtable<Key, Value, Hash, KeyEqual, Grower, Allocator>(args);
} 

````

I first write a factory class looked like above. Then I found a way to use a Registrar to registrar so I don't need to add switch branch in the future. 

I don't know which one is better, what is your opinion?

Reply:
> The purpose of HashTableFactory is to hide all hashtable's constructor into this class. The main difference for our hashtable and velox's hashtable is that we have many and they have only one. I guess you are thinking making this class looks like this
> 
> ```
> template <typename Key,
>           typename Value,
>           typename Hash,
>           typename KeyEqual,
>           typename Grower,
>           typename Allocator>
> class hashtablefactory 
> 
> template <typename... Args>
> BaseHashTable<Key, Value, Hash, KeyEqual, Grower, Allocator> createForJoin (HastTableType hashtable_type, Args.. args) {
>   switch(hashtable_type)
>       case xxx:
>           return new xx_hashtable<Key, Value, Hash, KeyEqual, Grower, Allocator>(args);
>       case yyy:
>           return new yy_hashtable<Key, Value, Hash, KeyEqual, Grower, Allocator>(args);
> } 
> ```
> 
> I first write a factory class looked like above. Then I found a way to use a Registrar to registrar so I don't need to add switch branch in the future.
> 
> I don't know which one is better, what is your opinion?

Inside the 'createForJoin' method , how about to get the hashTable from a pre-registered std::unordered_map<HashTableType,BaseHashTable>?